agent/src/main/scala/za/co/absa/atum/agent/model/MeasurementBuilder.scala

```diff
@@ -16,19 +16,40 @@
 
 package za.co.absa.atum.agent.model
 
+import za.co.absa.atum.agent.exception.AtumAgentException.MeasurementException
 import za.co.absa.atum.model.dto.{MeasureDTO, MeasureResultDTO, MeasurementDTO}
-import za.co.absa.atum.model.dto.MeasureResultDTO.TypedValue
 
 /**
  * This object provides a functionality to convert a measurement to its DTO representation.
  */
 private [agent] object MeasurementBuilder {
 
-  private [agent] def buildMeasurementDTO(measurement: Measurement): MeasurementDTO = {
-    val measureName = measurement.measure.measureName
-    val measureDTO = MeasureDTO(measureName, measurement.measure.controlColumns)
-    val measureResultDTO = MeasureResultDTO(TypedValue(measurement.resultValue.toString, measurement.resultType))
+  private[agent] def validateMeasureUniqueness(measurements: Set[Measurement]): Unit = {
+    val allMeasures = measurements.toSeq.map(_.measure)
 
+    val originalMeasuresCnt = allMeasures.size
+    val uniqueMeasureColumnCombinationCnt = allMeasures.map(m =>
+      Tuple2(m.measureName, m.measuredColumns)  // there can't be two same measures defined on the same column(s)
```

Comment:
The `Tuple2` is not needed.

Reply:
Maybe it could be simplified ...

```scala
if (measurements.size != measurements.toSeq.distinctBy(m => (m.measure.measureName, m.measure.measuredColumns)).size) {
      throw MeasurementException(
        s"Measure and measuredColumn combinations must be unique, i.e. they cannot repeat!"
      )
    }
```

Reply:
Did you think of any other approaches? `Set` uses `equals` and `hashCode` methods to determine uniqueness. You could overwrite those on `AtumMeasure` and no validation method would be needed.

Reply:
Hmm I was thinking about moving this code more to `AtumMeasure`, but in the end, the Set that the user specifies is not a set of AtumMeasures, but a Set of `Measurement` objects. So I thought that the validation method is warranted - and it's relatively 'fail-early' approach, since this would be executed during CheckpointDTO creation. But I still like your idea about using `distinctBy` method :)

Reply:
On the other hand it might make sense, technically speaking, to implement something like `equals` and `hashCode` directly on `Measurement`, but I feel like a direct, dedicated validation, that's extracted elsewhere, thus not changing the 'default' behavior of case class uniqueness in this case, seems like a better idea

Reply:
Hmm actually `distinctBy` is not generally available in Scala 2.12 IIRC - I'll still think about some improvements for sure

Reply:
Interesting, I had no idea it's 2.13 addition. I agree with your point about not changing the default behaviour.